core/src/scene/sceneLoader.cpp

```diff
@@ -965,66 +980,77 @@ void SceneLoader::parseStyleParams(Node params, Scene& scene, const std::string&
-            if (texItr == scene.textures().end()) {
-                loadTexture(strVal, scene);
+            styleUniform.type = "sampler2D";
+            std::shared_ptr<Texture> texture;
+            scene.texture(strVal, texture);
+
+            if (!texture && !loadTexture(strVal, scene)) {
+                LOGW("Can't load texture with name %s", strVal.c_str());
             }
-            uniformValues.push_back(strVal);
-        }
 
+            styleUniform.value = strVal;
+        }
     } else if (value.IsSequence()) {
         int size = value.size();
         try {
-            type = "vec" + std::to_string(size);
-            UniformValue uVal;
             switch (size) {
                 case 2:
-                    uVal = parseVec<glm::vec2>(value);
+                    styleUniform.value = parseVec<glm::vec2>(value);
                     break;
                 case 3:
-                    uVal = parseVec<glm::vec3>(value);
+                    styleUniform.value = parseVec<glm::vec3>(value);
                     break;
                 case 4:
-                    uVal = parseVec<glm::vec4>(value);
+                    styleUniform.value = parseVec<glm::vec4>(value);
                     break;
                 default:
+                    UniformArray uniformArray;
+                    for (const auto& val : value) {
+                        uniformArray.push_back(val.as<double>());
+                    }
+                    styleUniform.value = uniformArray;
                     break;
-                    // TODO: Handle numeric arrays past 4 elements
             }
-            uniformValues.push_back(uVal);
+
+            styleUniform.type = "vec" + std::to_string(size);
         } catch (const BadConversion& e) { // array of strings (textures)
-            uniformValues.reserve(size);
-            type = "sampler2D";
+            UniformTextureArray textureArrayUniform;
+            textureArrayUniform.names.reserve(size);
+            styleUniform.type = "sampler2D";
+
             for (const auto& strVal : value) {
                 auto textureName = strVal.as<std::string>();
-                uniformValues.push_back(textureName);
-                auto texItr = scene.textures().find(textureName);
-                if (texItr == scene.textures().end()) {
-                    loadTexture(textureName, scene);
+                textureArrayUniform.names.push_back(textureName);
+                std::shared_ptr<Texture> texture;
+                scene.texture(textureName, texture);
+
+                if (!texture && !loadTexture(textureName, scene)) {
+                    LOGW("Can't load texture with name %s", textureName.c_str());
                 }
             }
+
+            styleUniform.value = textureArrayUniform;
         }
     } else {
         LOGW("Expected a scalar or sequence value for uniforms");
     }
-    return std::make_pair(type, std::move(uniformValues));
```

Comment:
I think it's more clear to make the struct explicit, I found it a bit sketchy to use the pair as an implicit struct while refactoring this (mainly for readability), since you always have to look at what is `first`/`second` in the pair.